core/consensus/babe/impl/babe_synchronizer_impl.cpp

```diff
@@ -3,138 +3,803 @@
+                applying_res.error().message());
+            if (handler) handler(Error::DISCARDED_BLOCK);
+          } else {
+            SL_DEBUG(log_,
+                     "Block #{} hash={} is skipped as existing",
+                     block.header->number,
+                     hash.toHex());
+            if (handler) handler(applying_res.as_failure());
+          }
+        } else {
+          if (handler) handler(grandpa::BlockInfo(block.header->number, hash));
+        }
+      }
+    }
+    ancestry_.erase(hash);
+
+    if (known_blocks_.size() < 250) {
+      SL_TRACE(log_,
+               "{} blocks in queue: ask next portion of block",
+               known_blocks_.size());
+      askNextPortionOfBlocks();
+    } else {
+      SL_TRACE(log_, "{} blocks in queue", known_blocks_.size());
+    }
+
+    scheduler_->schedule([wp = weak_from_this()] {
+      if (auto self = wp.lock()) {
+        self->applyNextBlock();
+      }
+    });
+  }
+
+  size_t BabeSynchronizerImpl::discardBlock(
+      const primitives::BlockHash &hash_of_discarding_block) {
+    std::queue<primitives::BlockHash> queue;
+    queue.emplace(hash_of_discarding_block);
+
+    size_t affected = 0;
+    while (not queue.empty()) {
+      const auto &hash = queue.front();
+
+      affected += known_blocks_.erase(hash);
+
+      auto range = ancestry_.equal_range(hash);
+      for (auto it = range.first; it != range.second; ++it) {
+        queue.emplace(it->second);
+      }
+      ancestry_.erase(range.first, range.second);
+
+      queue.pop();
+    }
+
+    return affected;
+  }
+
+  void BabeSynchronizerImpl::prune(
+      const primitives::BlockInfo &finalized_block) {
+    // Remove blocks whose numbers less finalized one
+    while (not generations_.empty()) {
+      auto generation_node = generations_.extract(generations_.begin());
+      if (generation_node) {
+        const auto &number = generation_node.key();
+        if (number >= finalized_block.number) {
+          break;
+        }
+        const auto &hash = generation_node.mapped();
+        known_blocks_.erase(hash);
+        ancestry_.erase(hash);
+      }
+    }
+
+    // Remove blocks whose numbers equal finalized one, exceptly finalized one
+    auto range = generations_.equal_range(finalized_block.number);
+    for (auto it = range.first; it != range.second;) {
+      auto cit = it++;
+      const auto &hash = cit->second;
+      if (hash != finalized_block.hash) {
+        discardBlock(hash);
+      }
+    }
+  }
+
+  void BabeSynchronizerImpl::askNextPortionOfBlocks() {
+    bool false_val = false;
+    if (not asking_blocks_portion_in_progress_.compare_exchange_strong(
+            false_val, true)) {
+      SL_TRACE(log_, "Asking portion of blocks in progress");
+      return;
+    }
+    SL_TRACE(log_, "Begin asking portion of blocks");
+
+    for (auto g_it = generations_.rbegin(); g_it != generations_.rend();
+         ++g_it) {
```

Comment:
Hm, why doesn't for-each fit here? Are generations_ changed inside this loop?

Reply:
Reverse iterator is used here